accepted/cranelift-roadmap-2022.md

```diff
@@ -0,0 +1,895 @@
+# Cranelift Roadmap for 2022
+
+Around this time last year, we [published a roadmap for Cranelift in
+2021](https://github.com/bytecodealliance/rfcs/pull/8). It has been a
+productive year, and Cranelift is in a very different spot today than
+it was at the end of 2020. (A separate "progress report" on 2021 is
+forthcoming!) With all our work in the past year, we are now starting
+to have what might be called a "solid foundation": we have worked to
+improve many core parts of the compiler (e.g., ISLE, regalloc2), we
+have paid down technical debt by completing a backend transition and
+removing old code, we have significantly improved our backends
+(especially SIMD) and our fuzzing/testing situation (differential
+testing, the interpreter, custom mutator), and more. We will certainly
+continue to improve these fundamentals, but one might also reasonably
+ask: what are the next big steps we hope to take?
+
+This document continues the yearly-roadmap tradition by documenting
+our aspirations for 2022: general project directions, and specific
+milestones, that we hope to reach.
+
+## High-level Priorities
+
+In a single sentence, the priorities and direction for Cranelift in
+2022 can be summarized as: *improve the optimizer, make verification
+efforts concrete, expand the feature set as needed, and let ongoing
+work stabilize and mature*.
+
+In a little more detail, we identify the following five areas of work,
+which we will expand below:
+
+1. Complete the *ISLE*
+   ([RFC](https://github.com/bytecodealliance/rfcs/pull/15),
+   [compiler](https://github.com/bytecodealliance/wasmtime/tree/main/cranelift/isle/))
+   and *[regalloc2](https://github.com/bytecodealliance/regalloc2)*
+   efforts. These two projects improve the core compiler backend's
+   generated-code quality, maintainability, and performance, and
+   provide a solid foundation moving forward.
+   
+2. *Improve the compiler's [middle
+   end](https://en.wikipedia.org/wiki/Compiler#Middle_end)*, or
+   collection of IR-level optimizations. Cranelift has the basic
+   optimizations such as constant propagation/folding,
+   common-subexpression elimination, and dead-code elimination, but
+   some of these are fairly basic, and other common optimizations
+   (e.g., inlining) are missing; this will become more important in
+   some expected future use-cases. As we do this, we should look for
+   ways to facilitate verification efforts as well.
+   
+3. Add compiler support for *tail calls*, *exception handling*, and
```

Comment:
Should we mention the flexible vectors proposal (given that @sparker-arm is looking at it)?

Reply:
Added, thanks!